pygsti/extras/ibmq/ibmqcore.py

```diff
@@ -109,74 +78,123 @@ def __init__(self, edesign, pspec, remove_duplicates=True, randomized_order=True
         # Populated when grabbing results from IBM Q with .retrieve_results()
-        self['batch_result_object'] = None
-        self['data'] = None
-
-        circuits = edesign.all_circuits_needing_data.copy()
-        if randomized_order:
-            if remove_duplicates:
-                circuits = list(set(circuits))
-            _np.random.shuffle(circuits)
+        self.batch_results = []
+        self.data = None
+
+        # If not in this list, will be automatically dumped to meta.json
+        # 'none' means it will not be read in, 'reset' means it will come back in as None
+        # Several of these could be stored in the meta.json but are kept external for easy chkpts
+        self.auxfile_types['edesign'] = 'none'
+        self.auxfile_types['data'] = 'reset'
+        # self.processor_spec is handled by _HasPSpec base class
+        self.auxfile_types['pygsti_circuit_batches'] = 'list:text-circuit-list'
+        self.auxfile_types['qasm_circuit_batches'] = 'list:json'
+        self.auxfile_types['qiskit_circuit_batches'] = 'none'
+        self.auxfile_types['qjobs'] = 'none'
+        self.auxfile_types['job_ids'] = 'json'
+        if _json_util is not None:
+            self.auxfile_types['submit_time_calibration_data'] = 'list:json'
+            self.auxfile_types['batch_results'] = 'list:json'
         else:
-            assert(not remove_duplicates), "Can only remove duplicates if randomizing order!"
-
-        num_batches = int(_np.ceil(len(circuits) / circuits_per_batch))
-
-        self['pygsti_circuits'] = [[] for i in range(num_batches)]
-        self['pygsti_openqasm_circuits'] = [[] for i in range(num_batches)]
-        self['qiskit_QuantumCircuits'] = [[] for i in range(num_batches)]
-        self['qiskit_QuantumCircuits_as_openqasm'] = [[] for i in range(num_batches)]
-        self['submit_time_calibration_data'] = []
-        self['qobj'] = []
-
-        batch_idx = 0
-        for circ_idx, circ in enumerate(circuits):
-            self['pygsti_circuits'][batch_idx].append(circ)
-            if len(self['pygsti_circuits'][batch_idx]) == circuits_per_batch:
-                batch_idx += 1
-
-        #create Qiskit quantum circuit for each circuit defined in experiment list
-        total_num = 0
-
-        #start = _time.time()
-        for batch_idx, circuit_batch in enumerate(self['pygsti_circuits']):
-            print("Constructing job for circuit batch {} of {}".format(batch_idx + 1, num_batches))
-            #openqasm_circuit_ids = []
-            for circ_idx, circ in enumerate(circuit_batch):
-                pygsti_openqasm_circ = circ.convert_to_openqasm(num_qubits=pspec.num_qubits,
+            # Fall back to pickles if we do not have bson to deal with datetime.datetime
+            self.auxfile_types['submit_time_calibration_data'] = 'pickle'
+            self.auxfile_types['batch_results'] = 'pickle'
+        
+        if not disable_checkpointing:
+            if self.checkpoint_path is None:
```

Comment:
Since we're enabling checkpointing by default I think it would be worthwhile implementing a default path here for when the checkpoint_path kwarg isn't set by a user. This something we currently do in the GST protocol checkpointing code, fwiw, in case referencing how it is handled in that part of the code would be useful.

Reply:
I guess that is fair. One reason I did it this way is that the "checkpoint" here is actually a completely valid IBMQExperiment object that can be loaded using from_dir(), so in my experience so far it has been nice to have it match the dirname for write() so that the user doesn't have to change how they load from checkpoint vs a .write()... but it's a good point that a default path could simplify this for users. I'll think about how I want to incorporate that here.